sanic/server/websockets/impl.py

```diff
@@ -0,0 +1,776 @@
+                # connection_lost() could run between the moment the timeout occurs
+                # and the moment this coroutine resumes running.
+                return self.connection_lost_waiter.done()
+
+    async def process_events(self, events: Sequence[Event]) -> None:
+        """
+        Process a list of incoming events.
+        """
+        # Wrapped in a mutex lock, to prevent other incoming events
+        # from processing at the same time
+        async with self.process_event_mutex:
+            for event in events:
+                if event.opcode == OP_PONG:
+                    await self.process_pong(event)
+                else:
+                    await self.assembler.put(event)
+
+    async def process_pong(self, frame: "Frame") -> None:
+        if frame.data in self.pings:
+            # Acknowledge all pings up to the one matching this pong.
+            ping_ids = []
+            for ping_id, ping in self.pings.items():
+                ping_ids.append(ping_id)
+                if not ping.done():
+                    ping.set_result(None)
+                if ping_id == frame.data:
+                    break
+            else:  # noqa
+                raise ServerError("ping_id is not in self.pings")
+            # Remove acknowledged pings from self.pings.
+            for ping_id in ping_ids:
+                del self.pings[ping_id]
+
+    async def keepalive_ping(self) -> None:
+        """
+        Send a Ping frame and wait for a Pong frame at regular intervals.
+        This coroutine exits when the connection terminates and one of the
+        following happens:
+        - :meth:`ping` raises :exc:`ConnectionClosed`, or
+        - :meth:`auto_close_connection` cancels :attr:`keepalive_ping_task`.
+        """
+        if self.ping_interval is None:
+            return
+
+        try:
+            while True:
+                await asyncio.sleep(self.ping_interval)
+
+                # ping() raises CancelledError if the connection is closed,
+                # when auto_close_connection() cancels self.keepalive_ping_task.
+
+                # ping() raises ConnectionClosed if the connection is lost,
+                # when connection_lost() calls abort_pings().
+
+                ping_waiter = await self.ping()
+
+                if self.ping_timeout is not None:
+                    try:
+                        await asyncio.wait_for(ping_waiter, self.ping_timeout)
+                    except asyncio.TimeoutError:
+                        error_logger.warning(
+                            "Websocket timed out waiting for pong"
+                        )
+                        self.fail_connection(1011)
+                        break
+        except asyncio.CancelledError:
+            raise
+        except ConnectionClosed:
+            pass
+        except Exception:
+            error_logger.warning("Unexpected exception in keepalive ping task")
```

Comment:
We should probably log out what the exception is instead of swallowing it.

Reply:
Good idea.